src/Engine/Renderer/Apis/OpenGL/Shader/OpenGLMaterialData.cpp

```diff
@@ -1,168 +1,304 @@
 #include "Engine/Renderer/Shader/Shader.h"
+#include "Engine/Renderer/Texture/Texture2D.h"
 
 #include <glm.hpp>
 
-namespace gp1 {
-
-	OpenGLMaterialData::OpenGLMaterialData(Material* material)
-		: MaterialData(material) {}
+namespace gp1::renderer::apis::opengl::shader
+{
+	OpenGLMaterialData::OpenGLMaterialData(renderer::shader::Material* material)
+	    : OpenGLRendererData(material) {}
 
 	void OpenGLMaterialData::CleanUp() {}
 
-	RendererType OpenGLMaterialData::GetRendererType() const {
-		return RendererType::OPENGL;
+	GLenum OpenGLMaterialData::GetCullFace() const
+	{
+		return GetGLCullFace(GetDataUnsafe<renderer::shader::Material>()->m_CullMode.m_Face);
 	}
 
-	GLenum OpenGLMaterialData::GetCullFace() const {
-		return GetGLCullFace(GetMaterial<Material>()->m_CullMode.m_Face);
+	GLenum OpenGLMaterialData::GetSrcBlendFunc() const
+	{
+		return GetGLBlendFunc(GetDataUnsafe<renderer::shader::Material>()->m_BlendFunc.m_SrcFunc);
 	}
 
-	GLenum OpenGLMaterialData::GetSrcBlendFunc() const {
-		return GetGLBlendFunc(GetMaterial<Material>()->m_BlendFunc.m_SrcFunc);
+	GLenum OpenGLMaterialData::GetDstBlendFunc() const
+	{
+		return GetGLBlendFunc(GetDataUnsafe<renderer::shader::Material>()->m_BlendFunc.m_DstFunc);
 	}
 
-	GLenum OpenGLMaterialData::GetDstBlendFunc() const {
-		return GetGLBlendFunc(GetMaterial<Material>()->m_BlendFunc.m_DstFunc);
+	GLenum OpenGLMaterialData::GetPolygonModeFace() const
+	{
+		return GetGLCullFace(GetDataUnsafe<renderer::shader::Material>()->m_PolygonMode.m_Face);
 	}
 
-	GLenum OpenGLMaterialData::GetPolygonModeFace() const {
-		return GetGLCullFace(GetMaterial<Material>()->m_PolygonMode.m_Face);
+	GLenum OpenGLMaterialData::GetPolygonMode() const
+	{
+		return GetGLPolygonMode(GetDataUnsafe<renderer::shader::Material>()->m_PolygonMode.m_Mode);
 	}
 
-	GLenum OpenGLMaterialData::GetPolygonMode() const {
-		return GetGLPolygonMode(GetMaterial<Material>()->m_PolygonMode.m_Mode);
-	}
+	void OpenGLMaterialData::SetAllUniforms(OpenGLRenderer* renderer)
+	{
+		renderer::shader::Material* material = GetDataUnsafe<renderer::shader::Material>();
+		if (!material) return;
+		if (!material->GetShader()) return;
 
-	void OpenGLMaterialData::SetAllUniforms() {
-		if (!GetMaterial<Material>()->GetShader())
-			return;
+		uint32_t texIndex = 0;
+		for (auto& uniform : material->GetUniforms())
+		{
+			using namespace renderer::shader;
 
-		for (auto& uniform : GetUniforms()) {
-			const std::string& id = uniform.first;
-			uint32_t location = GetMaterial<Material>()->GetShader()->GetUniformLocation(id);
-			if (uniform.second.type() == typeid(Uniform<float>)) {
-				const Uniform<float>* uniformFloat = std::any_cast<Uniform<float>>(&uniform.second);
+			const std::string& id       = uniform.first;
+			uint32_t           location = material->GetShader()->GetUniformLocation(id);
+			if (uniform.second.type() == typeid(Uniform<float>))
+			{
+				const Uniform<float>* uniformFloat = uniform.second._Cast<Uniform<float>>();
```

Comment:
What is `_Cast`? If that is an MSVC-specific function then it's a big no-no.

Reply:
Isn't it weird how only MSVC has functions that make it look nicer.
But yeah I just tested and the `_Cast` is sadly an MSVC-specific function, so I'll get to swapping them back to `std::any_cast`